[Sema] Diagnose by-value copy constructors in template instantiations

Megan0704-1 opened the pull request:
Fixes #80963 

This PR ensures Clang diagnoses by-value copy constructors in implicitly instantiated class templates (e.g., `A<int, int>(A<int, int>)`), per [class.copy.ctor]. 

Changes: 
- Remove `TSK_ImplicitInstantiation` check in `SemaDeclCXX.cpp`.
- Add `!isFunctionTemplateSpecialization()` to skip templated constructors.
- Add regression tests.

Megan0704-1 committed:
[Sema] Diagnose by-value copy constructors in template instantiations

Fixes #80963

Previously, Clang skipped diagnosing a constructor if it was implicitly
instantiated from a template class (TSK_ImplicitInstantiation). This allowed
ill-formed “copy” constructors taking the class by value (e.g. A(A)) to slip
through without a diagnostic.

However, the C++ standard mandates that copy constructors must take their
class type parameter by reference (e.g., A(const A&)). Furthermore, a
constructor template that *would* form a copy-by-value signature is not
treated as a copy constructor and should never be chosen for copying.

This patch replaces the check on TSK_ImplicitInstantiation with a check
to see if the constructor is a function template specialization (i.e.,
isFunctionTemplateSpecialization()). That ensures proper diagnosis of
non-template copy-by-value constructors, while still allowing valid
template constructors that might appear to have a copy-like signature
but should be SFINAEd out or simply not selected as a copy constructor.

Megan0704-1 commented:
@erichkeane, @shafik 
Could you please take a look and let me know if it looks good to you? Thank you very much!

cor3ntin commented on clang/test/SemaCXX/copy-ctor-template.cpp:
```diff
@@ -0,0 +1,22 @@
+
+void g() {
+    B<int, int> b = B<int, int>(); // should use implicit copy constructor
+}
```
Can you add a test for @hubert-reinterpretcast comment here
https://github.com/llvm/llvm-project/issues/80963#issuecomment-2716322960

Can you add tests for something like
`A<int, double> a = A<double, int>();` // not a copy constructor

Megan0704-1 replied:
Thanks for the review! I’ll add tests for the A<int, double> case and update the PR shortly.

cor3ntin reviewed (commented):
Thanks for working on this 

Can you add a changelog entry in clang/docs/ReleaseNotes.rst (in the bug fixes to c++ section)?

Thanks

Megan0704-1 committed:
[Docs] Add release note for #80963 fix

Megan0704-1 committed:
[Sema] Fix test diagnostics for copy-ctor-template

Megan0704-1 commented:
Thanks for your feedback and review!! I’ve updated the changelog and the test file to include the rvalue-to-lvalue case and templated constructor scenario. Please let me know if there’s anything else!

cor3ntin reviewed (approved):
LGTM. Thanks for fixing this issue!

Will you need me to merge this for you?

Megan0704-1 commented:
Thank you for reviewing and approving the PR!
Yes, please merge it for me. Thanks again for all your help!

cor3ntin commented:
@Megan0704-1 Can you look into the test failures? thanks

Megan0704-1 committed:
[SemaCXX] Update constructor-template.cpp test for #80963

Adjust the test to account for new diagnostics emitted when a class template constructor takes the class by value after instantiation.
- Add expected-error annotations to lines 142, 159, 172 (invalid by-value constructors when T = U).
- Add expected-note annotations to lines 147, 164, 176 (template instantiation points).
- Remove expected-error from line 76 (valid template specialization).

Megan0704-1 commented:
@cor3ntin 
Thank you for flagging the test failures! 

I’ve investigated them, and here’s what I found: 
Failed Tests in constructor-template.cpp: 
- The failures occurred because the test wasn’t annotated to expect the new diagnostics introduced by the fix (e.g., A<int, int> instantiations now emit errors for invalid by-value copy constructors). 

Fix it by updating the test to: 
- Add expected-error to lines 142, 159, 172 (invalid by-value constructors when T = U). 
- Add expected-note to lines 147, 164, 176 (template instantiation points). 
- Remove an incorrect expected-error from line 76 (valid template specialization). 

Please Let me know if further adjustments are needed~

shafik commented on clang/lib/Sema/SemaDeclCXX.cpp:
```diff
@@ -10921,8 +10921,8 @@ void Sema::CheckConstructor(CXXConstructorDecl *Constructor) {
-      Constructor->getTemplateSpecializationKind() !=
-          TSK_ImplicitInstantiation) {
+      !Constructor->isFunctionTemplateSpecialization()
+          ) {
```
Did you run clang-format on this? I am surprised it would leave the `)` on the line like that but maybe I am off.

Megan0704-1 replied:
Thank you for catching this, and I am so sorry for the oversight.
I forgot to run clang-format before merging. 
Should I submitt a follow-up PR to fix the formatting issues?

avikivity commented:
> This error is [produced for the following code](https://github.com/carbon-language/carbon-lang/pull/5170):
> 
> ```c++
> template <typename KeyT>
> class SetView {
>   SetView(SetView<std::remove_const_t<KeyT>> other_view)
>     requires(!std::same_as<KeyT, std::remove_const_t<KeyT>>);
> };
> ```
> 
> ... which is never an eligible copy constructor. Is that intended? If this is in line with the language rules, perhaps we should file a core issue.
> 
> [Edit: I've mailed the core reflector.]

Can you link to the defect report please?
